File: Tree/C++/path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        vector<int> path;
        if (root == NULL) {
            return res;
        }
        helper(root, path, sum, res);
        return res;
    }
    
    void helper(TreeNode *root, vector<int>& path, int sum, vector<vector<int>>& res) {
        if (root == NULL) {
            return;
        }
        path.push_back(root->val);
        sum -= root->val;
        if (root->left == NULL && root->right == NULL) {
            if (sum == 0) {
                res.push_back(path);
            }
        } else {
            if (root->left) {
                helper(root->left, path, sum, res);
            }
            if (root->right) {
                helper(root->right, path, sum, res);
            }
        }
        path.pop_back();
        sum += root->val;
    }
};

File: Google/Golang/path-sum-ii.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pathSum(root *TreeNode, sum int) [][]int {
    res := make([][]int, 0)
	if root == nil {
		return res
	}
	path := make([]int, 0)
	dfs(root, sum, path, &res)
	return res
}

func dfs(root *TreeNode, sum int, path []int, res *[][]int) {
    if root == nil {
        return
    }
    path = append(path, root.Val)
    sum -= root.Val
    if root.Left == nil && root.Right == nil && sum == 0 {
        tmp := make([]int, len(path))
        copy(tmp, path)
		*res = append(*res, tmp)
    } else {
        if root.Left != nil {
		    dfs(root.Left, sum, path, res)
        }
	    if root.Right != nil {
		    dfs(root.Right, sum, path, res)
        }
	}
    path = path[:len(path)-1]
    sum += root.Val
}
